File: jupyter2/ps/parametric/generic_optimize/ml/generic_optimize.ml
(** begin my answer *)

(** begin hidden *)
let arange a b n =
  let dx = (b -. a) /. float_of_int (n - 1) in
  object 
    val mutable i = 0
    method next =
      if i = n then None
      else
        let x = a +. dx *. float_of_int i in
        let _ = i <- i + 1 in
        Some x
  end

let arange2d x0 x1 m y0 y1 n =
  let dx = (x1 -. x0) /. float_of_int (m - 1) in
  let dy = (y1 -. y0) /. float_of_int (n - 1) in
  object 
    val mutable i = 0
    val mutable j = 0
    method next =
      if i = m then
        None
      else 
        let x = x0 +. dx *. float_of_int i in
        let y = y0 +. dy *. float_of_int j in
        let _ = i <- if j + 1 = n then i + 1 else i in
        let _ = j <- if j + 1 = n then 0 else j + 1 in
        Some (x, y)
  end

let optimize f rng =
  let rec iter x miny =
    match (x, miny) with
    | (None, _) -> miny
    | (Some x, None) ->
       iter rng#next (Some(f x))
    | (Some x, Some(miny)) ->
       let y = f x in
       let next = rng#next in
       if y < miny then
         iter next (Some(y))
       else
         iter next (Some(miny))
  in
  iter (rng#next) None
(** end hidden *)
(** end my answer *)

let very_close a b =
  match a with
  | None -> false
  | Some a -> Float.abs (a -. b) < 1e-6

let main () =
  let f0 x = x *. (x -. 1.) in
  let f1 x = x *. (x -. 1.) *. (x -. 2.) in
  let f2 (x, y) = x *. x +. 2. *. y *. y +. 3. *. x *. y in
  let a0 = arange (-1.) 3. 10000 in
  let a1 = arange (-0.) 3. 10000 in
  let a2 = arange2d (-2.) 2. 10000 (-2.) 2. 1000 in
  let y0 = optimize f0 a0 in
  let y1 = optimize f1 a1 in
  let y2 = optimize f2 a2 in
  assert (very_close y0 (-0.25));
  assert (very_close y1 (-0.384900));
  assert (very_close y2 (-0.5));

  Printf.printf "OK\n"
;;

main()

File: jupyter2/ps/parametric/generic_optimize/rs/generic_optimize.rs
/** begin my answer */

/** begin hidden */
trait Domain<T> {
    fn next(self : &mut Self) -> Option<T>;
}

#[allow(dead_code)]
struct Arange {
    a : f64,
    b : f64,
    i : i64,
    n : i64,
    dx : f64
}

fn arange(a : f64, b : f64, n : i64) -> Arange {
    Arange{a, b, i: 0, n, dx: (b - a) / ((n - 1) as f64)}
}

impl Domain<f64> for Arange {
    fn next(self : &mut Arange) -> Option<f64> {
	if self.i == self.n {
	    None
	} else {
	    let x = self.a + self.dx * (self.i as f64);
	    self.i += 1;
	    Some(x)
	}
    }
}

#[allow(dead_code)]
struct Arange2d {
    x0 : f64,
    x1 : f64,
    m  : i64,
    y0 : f64,
    y1 : f64,
    n  : i64,
    i  : i64,
    j  : i64,
    dx : f64,
    dy : f64
}

fn arange2d(x0 : f64, x1 : f64, m : i64, y0 : f64, y1 : f64, n : i64) -> Arange2d {
    Arange2d{x0, x1, m, y0, y1, n, i: 0, j: 0, dx: (x1 - x0) / ((m - 1) as f64), dy: (y1 - y0) / ((n - 1) as f64)}
}

impl Domain<(f64, f64)> for Arange2d {
    fn next(self : &mut Arange2d) -> Option<(f64, f64)> {
	if self.i == self.m {
	    None
	} else {
	    let x = self.x0 + self.dx * (self.i as f64);
	    let y = self.y0 + self.dy * (self.j as f64);
	    if self.j + 1 == self.n {
		self.i += 1;
		self.j = 0;
	    } else {
		self.j += 1;
	    }
	    Some((x, y))
	}
    }
}

fn optimize<S : Clone>(f : fn(S) -> f64, dom : &mut dyn Domain<S>) -> Option<f64> {
    match dom.next() {
	None => None,
	Some(x) => {
	    let mut miny = f(x);
	    while let Some(x) = dom.next() {
		let y = f(x);
		if y < miny {
		    miny = y;
		}
	    }
	    Some(miny)
	}
    }
}
/** end hidden */
/** end my answer */

fn very_close(a : Option<f64>, b : f64) -> bool {
    match a {
	None => false,
	Some(a) => (a - b).abs() < 1e-6
    }
}

fn main() {
    fn f0(x : f64) -> f64 { x * (x - 1.) }
    fn f1(x : f64) -> f64 { x * (x - 1.) * (x - 2.) }
    fn f2((x, y) : (f64, f64)) -> f64 { x * x + 2. * y * y + 3. * x * y }
    let mut a0 = arange(0., 1., 10000);
    let mut a1 = arange(0., 3., 10000);
    let mut a2 = arange2d(-2., 2., 10000, -2., 2., 1000);
    let y0 = optimize(f0, &mut a0);
    let y1 = optimize(f1, &mut a1);
    let y2 = optimize(f2, &mut a2);
    very_close(y0, -0.25);
    very_close(y1, -0.384900);
    very_close(y2, -0.5);
    println!("OK")
}
